Caching allocator tracing

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #86241

We currently can take snapshots of the state of the allocated cuda memory, but we do not have a way to correlate these snapshots with the actions the allocator that were taken between snapshots. This PR adds a simple fixed-sized buffer that records the major actions that the allocator takes (ALLOC, FREE, SEGMENT_ALLOC, SEGMENT_FREE, OOM, SNAPSHOT) and includes these with the snapshot information. Capturing period snapshots with a big enough trace buffer makes it possible to see how the allocator state changes over time. 

We plan to use this functionality to guide how settings in the allocator can be adjusted and eventually have a more robust overall algorithm.

As a component of this functionality, we also add the ability to get a callback when the allocator will throw an OOM, primarily so that snapshots can be taken immediately to see why the program ran out of memory (most programs have some C++ state that would free tensors before the OutOfMemory exception can be caught).

This PR also updates the _memory_viz.py script to pretty-print the trace information and provide a better textual summary of snapshots distinguishing between internal and external fragmentation.

zdevito committed:
Caching allocator tracing

Adds a ring-buffer to the memory history that keeps a rolling buffer of
the last N actions that the allocator did.

* _memory_viz is updated to print better information about segments
* _memory_viz can pretty print the trace

TODO:

* Make the N configurable, so the amount of data in the snapshot can be limited
* Fix race condition problems with exposing internal caching allocator structures.

[ghstack-poisoned]

zdevito committed:
Update on "Caching allocator tracing"


Adds a ring-buffer to the memory history that keeps a rolling buffer of
the last N actions that the allocator did.

* _memory_viz is updated to print better information about segments
* _memory_viz can pretty print the trace

TODO:

* Make the N configurable, so the amount of data in the snapshot can be limited
* Fix race condition problems with exposing internal caching allocator structures.

[ghstack-poisoned]

zdevito committed:
Update on "Caching allocator tracing"


Adds a ring-buffer to the memory history that keeps a rolling buffer of
the last N actions that the allocator did.

* _memory_viz is updated to print better information about segments
* _memory_viz can pretty print the trace

TODO:

* Make the N configurable, so the amount of data in the snapshot can be limited
* Fix race condition problems with exposing internal caching allocator structures.

[ghstack-poisoned]

zdevito committed:
Update on "Caching allocator tracing"


Adds a ring-buffer to the memory history that keeps a rolling buffer of
the last N actions that the allocator did.

* _memory_viz is updated to print better information about segments
* _memory_viz can pretty print the trace

TODO:

* Make the N configurable, so the amount of data in the snapshot can be limited
* Fix race condition problems with exposing internal caching allocator structures.

[ghstack-poisoned]

zdevito committed:
Update on "Caching allocator tracing"


Adds a ring-buffer to the memory history that keeps a rolling buffer of
the last N actions that the allocator did.

* _memory_viz is updated to print better information about segments
* _memory_viz can pretty print the trace


[ghstack-poisoned]

zdevito committed:
Update on "Caching allocator tracing"


We currently can take snapshots of the state of the allocated cuda memory, but we do not have a way to correlate these snapshots with the actions the allocator that were taken between snapshots. This PR adds a simple fixed-sized buffer that records the major actions that the allocator takes (ALLOC, FREE, SEGMENT_ALLOC, SEGMENT_FREE, OOM, SNAPSHOT) and includes these with the snapshot information. Capturing period snapshots with a big enough trace buffer makes it possible to see how the allocator state changes over time. 

We plan to use this functionality to guide how settings in the allocator can be adjusted and eventually have a more robust overall algorithm.

As a component of this functionality, we also add the ability to get a callback when the allocator will throw an OOM, primarily so that snapshots can be taken immediately to see why the program ran out of memory (most programs have some C++ state that would free tensors before the OutOfMemory exception can be caught).

This PR also updates the _memory_viz.py script to pretty-print the trace information and provide a better textual summary of snapshots distinguishing between internal and external fragmentation.

[ghstack-poisoned]

ngimel commented on c10/cuda/CUDACachingAllocator.h:
```diff
@@ -131,6 +128,31 @@ struct SegmentInfo {
 };
 
+struct TraceEntry {
+  enum Action { ALLOC, FREE, SEGMENT_ALLOC, SEGMENT_FREE, SNAPSHOT, OOM };
```
can you leave small comments about what each of the actions mean?

ngimel commented on test/test_cuda.py:
```diff
@@ -4713,16 +4719,33 @@ def test_cpp_memory_snapshot_pickle(self):
+        x = False
+
+        def cb(device, alloc, device_alloc, device_free):
+            print(device, alloc, device_alloc, device_free)
```
debug print?

ngimel commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -804,6 +864,14 @@ class DeviceCachingAllocator {
     });
+    if (block->history) {
+      record_trace(
+          TraceEntry::FREE,
```
I'm uncertain if FREE should be recorded here - true, this is where FREE request happen, but actual freeing might not happen if there are other streams using this block (see the check below), so trace would be confusing - it would imply that memory was freed when in fact it wasn't.

ngimel commented on test/test_cuda.py:
```diff
@@ -4619,20 +4619,26 @@ def test_memory_snapshot(self):
+            del x
+            torch.cuda.empty_cache()
+            ss = torch.cuda.memory._snapshot()
+            assert(ss['device_traces'][0][-1]['action'] == 'segment_free')
```
self.assertTrue

zdevito committed:
Update on "Caching allocator tracing"


We currently can take snapshots of the state of the allocated cuda memory, but we do not have a way to correlate these snapshots with the actions the allocator that were taken between snapshots. This PR adds a simple fixed-sized buffer that records the major actions that the allocator takes (ALLOC, FREE, SEGMENT_ALLOC, SEGMENT_FREE, OOM, SNAPSHOT) and includes these with the snapshot information. Capturing period snapshots with a big enough trace buffer makes it possible to see how the allocator state changes over time. 

We plan to use this functionality to guide how settings in the allocator can be adjusted and eventually have a more robust overall algorithm.

As a component of this functionality, we also add the ability to get a callback when the allocator will throw an OOM, primarily so that snapshots can be taken immediately to see why the program ran out of memory (most programs have some C++ state that would free tensors before the OutOfMemory exception can be caught).

This PR also updates the _memory_viz.py script to pretty-print the trace information and provide a better textual summary of snapshots distinguishing between internal and external fragmentation.

[ghstack-poisoned]

zdevito committed:
Update on "Caching allocator tracing"


We currently can take snapshots of the state of the allocated cuda memory, but we do not have a way to correlate these snapshots with the actions the allocator that were taken between snapshots. This PR adds a simple fixed-sized buffer that records the major actions that the allocator takes (ALLOC, FREE, SEGMENT_ALLOC, SEGMENT_FREE, OOM, SNAPSHOT) and includes these with the snapshot information. Capturing period snapshots with a big enough trace buffer makes it possible to see how the allocator state changes over time. 

We plan to use this functionality to guide how settings in the allocator can be adjusted and eventually have a more robust overall algorithm.

As a component of this functionality, we also add the ability to get a callback when the allocator will throw an OOM, primarily so that snapshots can be taken immediately to see why the program ran out of memory (most programs have some C++ state that would free tensors before the OutOfMemory exception can be caught).

This PR also updates the _memory_viz.py script to pretty-print the trace information and provide a better textual summary of snapshots distinguishing between internal and external fragmentation.

[ghstack-poisoned]

zdevito commented:
@pytorchbot merge

zdevito commented:
@pytorchbot merge

zdevito commented:
@pytorchbot merge -f "tests appear to be passing but thinks a green check didn't pass"

zdevito committed:
Update on "Caching allocator tracing"


We currently can take snapshots of the state of the allocated cuda memory, but we do not have a way to correlate these snapshots with the actions the allocator that were taken between snapshots. This PR adds a simple fixed-sized buffer that records the major actions that the allocator takes (ALLOC, FREE, SEGMENT_ALLOC, SEGMENT_FREE, OOM, SNAPSHOT) and includes these with the snapshot information. Capturing period snapshots with a big enough trace buffer makes it possible to see how the allocator state changes over time. 

We plan to use this functionality to guide how settings in the allocator can be adjusted and eventually have a more robust overall algorithm.

As a component of this functionality, we also add the ability to get a callback when the allocator will throw an OOM, primarily so that snapshots can be taken immediately to see why the program ran out of memory (most programs have some C++ state that would free tensors before the OutOfMemory exception can be caught).

This PR also updates the _memory_viz.py script to pretty-print the trace information and provide a better textual summary of snapshots distinguishing between internal and external fragmentation.

[ghstack-poisoned]

zdevito commented:
@pytorchbot merge

zdevito committed:
Update on "Caching allocator tracing"


We currently can take snapshots of the state of the allocated cuda memory, but we do not have a way to correlate these snapshots with the actions the allocator that were taken between snapshots. This PR adds a simple fixed-sized buffer that records the major actions that the allocator takes (ALLOC, FREE, SEGMENT_ALLOC, SEGMENT_FREE, OOM, SNAPSHOT) and includes these with the snapshot information. Capturing period snapshots with a big enough trace buffer makes it possible to see how the allocator state changes over time. 

We plan to use this functionality to guide how settings in the allocator can be adjusted and eventually have a more robust overall algorithm.

As a component of this functionality, we also add the ability to get a callback when the allocator will throw an OOM, primarily so that snapshots can be taken immediately to see why the program ran out of memory (most programs have some C++ state that would free tensors before the OutOfMemory exception can be caught).

This PR also updates the _memory_viz.py script to pretty-print the trace information and provide a better textual summary of snapshots distinguishing between internal and external fragmentation.

[ghstack-poisoned]
